src/umpire/strategy/DynamicPoolMap.cpp

```diff
@@ -301,6 +234,129 @@ MemoryResourceTraits DynamicPoolMap::getTraits() const noexcept
+  if (m_free_map.size() > 1) {
+    const std::size_t released_bytes{releaseFreeBlocks()};
+    // Deallocated and removed released_bytes from m_free_map
+
+    // If this removed anything from the map, re-allocate a single large chunk
+    // and insert to free map
+    if (released_bytes > 0) {
+      UMPIRE_LOG(Debug, "coalescing " << released_bytes << " bytes.");
+      Pointer ptr{allocateBlock(released_bytes)};
+      insertFree(ptr, released_bytes, true, released_bytes);
+    }
+  }
+}
+
+void DynamicPoolMap::insertUsed(Pointer addr, std::size_t bytes, bool is_head,
+                                std::size_t whole_bytes)
+{
+  m_used_map.insert(
+      std::make_pair(addr, std::make_tuple(bytes, is_head, whole_bytes)));
+}
+
+void DynamicPoolMap::insertFree(Pointer addr, std::size_t bytes, bool is_head,
+                                std::size_t whole_bytes)
+{
+  m_free_map.insert(
+      std::make_pair(bytes, std::make_tuple(addr, is_head, whole_bytes)));
+}
+
+DynamicPoolMap::SizeMap::const_iterator DynamicPoolMap::findFreeBlock(
+    std::size_t bytes) const
+{
+  SizeMap::const_iterator iter{m_free_map.upper_bound(bytes)};
+
+  if (iter != m_free_map.begin()) {
+    // Back up iterator
+    --iter;
+    const std::size_t test_bytes{iter->first};
+    if (test_bytes < bytes) {
+      // Too small, reset iterator to what upper_bound returned
+      ++iter;
+    }
+  }
+
+  return iter;
+}
+
+void* DynamicPoolMap::allocateBlock(std::size_t bytes)
+{
+  UMPIRE_LOG(Debug, "Allocating new chunk of size " << bytes);
+
+  void* ptr{nullptr};
+  try {
+#if defined(UMPIRE_ENABLE_BACKTRACE)
+    {
+      umpire::util::backtrace bt;
+      umpire::util::backtracer<>::get_backtrace(bt);
+      UMPIRE_LOG(Info,
+                 "actual_size: " << (m_actual_bytes + bytes)
+                                 << " (prev: " << m_actual_bytes << ") "
+                                 << umpire::util::backtracer<>::print(bt));
+    }
+#endif
+    ptr = aligned_allocate(bytes);
+  } catch (...) {
+    UMPIRE_LOG(Error,
+                "Caught error allocating new chunk, giving up free chunks and "
+                "retrying...");
+    mergeFreeBlocks();
+    releaseFreeBlocks();
+    try {
+      ptr = aligned_allocate(bytes);
+      UMPIRE_LOG(Debug, "memory reclaimed, chunk successfully allocated.");
+    } catch (...) {
+      UMPIRE_LOG(Error, "recovery failed.");
+      throw;
+    }
+  }
+
+  UMPIRE_POISON_MEMORY_REGION(m_allocator, ptr, bytes);
+
+  m_actual_bytes += bytes;
+
+  return ptr;
+}
+
+void DynamicPoolMap::deallocateBlock(void* ptr, std::size_t size)
+{
+  UMPIRE_POISON_MEMORY_REGION(m_allocator, ptr, size);
+  m_actual_bytes -= size;
+  try {
+    aligned_deallocate(ptr);
+  }
+  catch (...) {
+    if (m_is_destructing) {
+      //
+      // Ignore the error if we are destructing as the resource to which we
+      // are releasing to may have gone away
+      //
+      UMPIRE_LOG(Error, "Pool is destructing, Exception Ignored");
```

Comment:
I think you should print the exception message. The underlying allocator will (and should) always exist. The comment above doesn't reflect the real issue this is fixing.

Reply:
I will fix the comment, but the underlying umpire exception is already printed right (by the call to UMPIRE_ERROR)

Reply:
Yup, you're correct.

Reply:
and sorry, but the comment is correct.  I will change it to: "...the resource to which we are releasing from...".  The 'resource' could be anything that offers a mechanism external to Umpire to cause it to be reset.

Reply:
Done